### PR TITLE
Add error dialog "Problem finding files"

### PR DESCRIPTION
**Pull request by IfIWantedTo**
I added an Error Dialog in AutoSetFileLinks. #5890 
I decided to open a pull request in order to get feedback on what I've done. 

This is my first contribution to OpenSource ever, so beware of stupid mistakes.

I am currently trying to go through the motions of contributing. However, I have run into a
few issues:

	 `ERROR org.jabref.logic.l10n.Localization - Messages are not initialized before accessing key: Your current Java version (%0) is not supported. Please install version %1 or higher. - `

- [ ] Change in CHANGELOG.md described (if applicable)
- [ ] Tests created for changes (if applicable)
- [ ] Manually tested changed features in running JabRef (always required)
- [ ] Screenshots added in PR description (for UI changes)
- [ ] [Checked documentation](https://docs.jabref.org/): Is the information available and up to date? If not created an issue at <https://github.com/JabRef/user-documentation/issues> or, even better, submitted a pull request to the documentation repository.

### CODE REVIEW

**Comment by Siedlerchr** [comment]
@IfIWantedTo  Welcome to JabRef!
I will try to give you some context.
Regarding the localization error, this seems your IDE configuration is not correct. Please check with the  guide here 
https://devdocs.jabref.org/getting-into-the-code/guidelines-for-setting-up-a-local-workspace#configure-your-ide

Back to the code, it's a bit more complicated then simply adding a dialog there.
The method has a loop so therefore a dialog would be a not a good idea for all files

I would propose to create a list with the exceptions in that class 
Then in the AutoLinkFilesAction in the succeed method  of the linksFilesTask you can check for that list if it has any exceptions  and report a notifications (the notify method shows a toast message that there were errors and say sth like go to the log for more details)

**Comment by IfIWantedTo** [comment]
@Siedlerchr Thank u for your help!

I have completely missed the loop before, thank you! I have replaced the code before with an iteration of
pushing to a list of exceptions as you suggested. I will still have to figure out where exactly to put the calling of notify() but I'm sure I will figure that out once I'm actually able to run the code.

Regarding the IDE: I figured there was a problem with the configuration and I went over the guide multiple times.
When it didn't seem to help I disregarded it. Now that you have confirmed my former suspicions I will resume my efforts. 

Again, thank you for your comment! Without you, this would have taken much longer :)

File: src/main/java/org/jabref/gui/externalfiles/AutoSetFileLinksUtil.java [hunk 1]
```diff
@@ -48,16 +52,24 @@ private AutoSetFileLinksUtil(List<Path> directories, AutoLinkPreferences autoLin
...
             try {
                 linkedFiles = findAssociatedNotLinkedFiles(entry);
             } catch (IOException e) {
+                fileExceptions.add(e); //change
                 LOGGER.error("Problem finding files", e);
+
             }
 
             if (ce != null) {
+                if (!fileExceptions.isEmpty()) { // is this where it goes? find out!
```
Comment (Siedlerchr) [inline src/main/java/org/jabref/gui/externalfiles/AutoSetFileLinksUtil.java:69]: I would propose to move this into the succeeded piece of code:
https://github.com/JabRef/jabref/blob/5ab494e8b1732f688d4ed8d696740bc7e8ef4101/src/main/java/org/jabref/gui/externalfiles/AutoLinkFilesAction.java#L60-L68

Comment (IfIWantedTo) [reply src/main/java/org/jabref/gui/externalfiles/AutoSetFileLinksUtil.java:69]: Oh ok, thank you. Can I ask you a few questions about the `succeeded()` method?

As far as I understood, it notifies the user when it has set the file links, i.e. when the returned List of linkAssociatedFiles(...) is not empty. In the other case `Finished automatically setting external links. No files found.`  I haven't quite understood yet, how this `else` case is different from what I am trying to accomplish.

Comment (IfIWantedTo) [reply src/main/java/org/jabref/gui/externalfiles/AutoSetFileLinksUtil.java]: Oh ok, thank you. Can I ask you a few questions about the `succeeded()` method?

As far as I understood, it notifies the user when it has set the file links, i.e. when the returned List of linkAssociatedFiles(...) is not empty. In the other case `Finished automatically setting external links. No files found.`  I haven't quite understood yet, how this `else` case is different from what I am trying to accomplish.

Comment (Siedlerchr) [reply src/main/java/org/jabref/gui/externalfiles/AutoSetFileLinksUtil.java]: Sorry for the delay in answering. The succeeded method is called whenever the task is finished normally and doesn't get interrupted. 
It either says files found or no files found, but it doesn't tell the user if there was an error (exception) which prevented  finding the files. 
So one addtional else case (your changes) is to notify the users that there have been X exceptions with the hint 
e.g. There were x files with an exception. See log for details

File: src/main/java/org/jabref/gui/externalfiles/AutoLinkFilesAction.java [hunk 2]
```diff
@@ -66,14 +66,19 @@ public void execute() {
...
+                        }
+                        dialogService.notify(Localization.lang("Finished automatically setting external links."));
+                    } else {
+                        dialogService.notify(Localization.lang("Finished automatically setting external links.") + " " + Localization.lang("No files found."));
                     }
-                    dialogService.notify(Localization.lang("Finished automatically setting external links."));
                 } else {
-                    dialogService.notify(Localization.lang("Finished automatically setting external links.") + " " + Localization.lang("No files found."));
+
+                    dialogService.notify(Localization.lang("Problem finding files")); //change
```
Comment (Siedlerchr) [inline src/main/java/org/jabref/gui/externalfiles/AutoLinkFilesAction.java:81]: You need to add another else case when the exceptions are not empty:
`Problem finding files.  Errors occured See log for details`

Comment (Siedlerchr) [reply src/main/java/org/jabref/gui/externalfiles/AutoLinkFilesAction.java:81]: And when no exception and getValue is empty => No files found

**Comment by calixtus** [comment]
I tried to finish this PR. Just did not had a file at hand to test the actual auto link, but every other case. So i think this should work, since I did not change the actual logic of the method. So ready for review.

File: src/main/java/org/jabref/gui/externalfiles/AutoLinkFilesAction.java [hunk 3]
```diff
@@ -34,47 +33,62 @@
...
+        final List<BibEntry> entries = stateManager.getSelectedEntries();
+        final AutoSetFileLinksUtil util = new AutoSetFileLinksUtil(
                 database,
                 preferences.getFilePreferences(),
                 preferences.getAutoLinkPreferences(),
                 ExternalFileTypes.getInstance());
-        Task<List<BibEntry>> linkFilesTask = new Task<>() {
+        final NamedCompound nc = new NamedCompound(Localization.lang("Automatically set file links"));
+
+        if (entries.stream().mapToInt(entry -> entry.getFiles().size()).sum() == 0) {
```
Comment (tobiasdiez) [inline src/main/java/org/jabref/gui/externalfiles/AutoLinkFilesAction.java:58]: `entries.stream().anyMatch(entry -> !entries.getFiles().isEmpty())` is probably easier to understand.

Comment (tobiasdiez) [reply src/main/java/org/jabref/gui/externalfiles/AutoLinkFilesAction.java:58]: But why is this check added in the first place? The autolinkfiles feature tries to find and link new files, so I don't understand why the entry already needs to have a file linked?

File: src/main/java/org/jabref/gui/externalfiles/AutoSetFileLinksUtil.java [hunk 4]
```diff
@@ -31,10 +31,31 @@
...
+
+        protected void addBibEntry(BibEntry bibEntry) {
+            changedEntries.add(bibEntry);
+        }
+
+        protected void addFileException(IOException exception) {
+            fileExceptions.add(exception);
+        }
+
+        public List<BibEntry> getBibEntries() {
```
Comment (tobiasdiez) [inline src/main/java/org/jabref/gui/externalfiles/AutoSetFileLinksUtil.java:46]: maybe rename to `getChangedEntries` ?

File: src/main/java/org/jabref/gui/externalfiles/AutoLinkFilesAction.java [hunk 5]
```diff
@@ -34,47 +33,62 @@
...
-                        nc.end();
-                        undoManager.addEdit(nc);
-                    }
-                    dialogService.notify(Localization.lang("Finished automatically setting external links."));
-                } else {
-                    dialogService.notify(Localization.lang("Finished automatically setting external links.") + " " + Localization.lang("No files found."));
+                AutoSetFileLinksUtil.LinkFilesResult result = getValue();
+
+                if (!result.getFileExceptions().isEmpty()) {
+                    dialogService.notify(Localization.lang("Problem finding files. See error log for details."));
```
Comment (tobiasdiez) [inline src/main/java/org/jabref/gui/externalfiles/AutoLinkFilesAction.java:74]: I think this should be a normal dialog. I would reserve `notify` for "everything went well"-messages.

File: src/main/java/org/jabref/gui/externalfiles/AutoLinkFilesAction.java [hunk 6]
```diff
@@ -34,47 +33,62 @@
...
+                    return;
+                }
+
+                if (nc.hasEdits()) {
+                    nc.end();
+                    undoManager.addEdit(nc);
+                }
+
+                dialogService.notify(Localization.lang(
+                        "Finished automatically setting external links. Changed %0 entries.",
```
Comment (tobiasdiez) [inline src/main/java/org/jabref/gui/externalfiles/AutoLinkFilesAction.java:90]: Reuse `Localization.lang("Finished automatically setting external links.")` from above by splitting the `Localization.lang(...)` here.

**Review by tobiasdiez — CHANGES_REQUESTED** [review]
Thanks @IfIWantedTo for your contribution and @calixtus for finishing this PR.

File: src/main/java/org/jabref/gui/externalfiles/AutoLinkFilesAction.java (hunk 3 above)
Comment (calixtus) [reply src/main/java/org/jabref/gui/externalfiles/AutoLinkFilesAction.java:58]: As I understand this feature, this calls AutoLinkFilesLinkUtil::findAssociatedNotLinkedFiles to check for broken file links. But those are also represented by a LinkedFile. So this feature tries to repair the links.

Comment (tobiasdiez) [reply src/main/java/org/jabref/gui/externalfiles/AutoLinkFilesAction.java:58]: It looks for files on disk that are matched by the citation key (= "associated", probably a bad name), but which are not yet have this file in the `file` field (= "linked"). https://github.com/JabRef/jabref/blob/a13f3075122256535dba7f8c32ca4f196b68856c/src/main/java/org/jabref/gui/externalfiles/AutoSetFileLinksUtil.java#L88-L92

So it's not about existing links that no longer point to a valid file (I think we have an integrity check for this though).

Comment (calixtus) [reply src/main/java/org/jabref/gui/externalfiles/AutoLinkFilesAction.java:58]: Oh ok. I totally misunderstood this. Sorry. I'll change that.

**Review by Siedlerchr — APPROVED** [review]

**Pull request merged**